crates/cli/src/main.rs

```diff
@@ -27,72 +31,90 @@ use futures::{
-
-        io.extend_with(NetNamespaceT::to_delegate(node.clone()));
-        io.extend_with(Web3NamespaceT::to_delegate(node.clone()));
-        io.extend_with(ConfigurationApiNamespaceT::to_delegate(node.clone()));
-        io.extend_with(DebugNamespaceT::to_delegate(node.clone()));
-        io.extend_with(EthNamespaceT::to_delegate(node.clone()));
-        io.extend_with(EthTestNodeNamespaceT::to_delegate(node.clone()));
-        io.extend_with(AnvilNamespaceT::to_delegate(node.clone()));
-        io.extend_with(EvmNamespaceT::to_delegate(node.clone()));
-        io.extend_with(HardhatNamespaceT::to_delegate(node.clone()));
-        io.extend_with(ZksNamespaceT::to_delegate(node));
-        io
-    };
-
-    std::thread::spawn(move || {
-        let runtime = tokio::runtime::Builder::new_multi_thread()
-            .enable_all()
-            .worker_threads(1)
-            .build()
-            .unwrap();
-
-        let allow_origin = if disable_cors {
-            "null"
+    let mut rpc = RpcModule::new(());
+    rpc.merge(EthNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(EthTestNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(AnvilNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(EvmNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(DebugNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(NetNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(ConfigNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(ZksNamespace::new(node.clone()).into_rpc())
+        .unwrap();
+    rpc.merge(Web3Namespace.into_rpc()).unwrap();
+
+    let cors_layers = tower::util::option_layer(if !disable_cors {
+        // `CorsLayer` adds CORS-specific headers to responses but does not do filtering by itself
+        let cors_layer = CorsLayer::new()
+            .allow_origin(AllowOrigin::exact(
+                cors_allow_origin.parse().expect("malformed allow origin"),
+            ))
+            .allow_headers([http::header::CONTENT_TYPE])
+            .allow_methods([Method::GET, Method::POST])
+            .expose_headers(Any);
+        // `HostFilterLayer` filters requests based on "Host" header
+        let host_filter_layer = if cors_allow_origin == "*" {
+            HostFilterLayer::disable()
         } else {
-            &cors_allow_origin
+            HostFilterLayer::new([cors_allow_origin]).expect("malformed allow origin")
         };
-        let mut builder = jsonrpc_http_server::ServerBuilder::new(io_handler)
-            .threads(1)
-            .event_loop_executor(runtime.handle().clone())
-            .cors(DomainsValidation::AllowOnly(vec![allow_origin.into()]));
+        Some((cors_layer, host_filter_layer))
+    } else {
+        None
+    });
+    let health_api_layer = tower::util::option_layer(if enable_health_api {
```

Comment:
Bikeshedding: Looks like one of the rare cases where [bool::then](https://doc.rust-lang.org/std/primitive.bool.html#method.then) would help with readability.